drivers/ethernet/eth_mcux.c

```diff
@@ -807,7 +812,11 @@ static void eth_rx(struct eth_context *context)
 
 	irq_unlock(imask);
 
-	if (net_recv_data(get_iface(context, vlan_tag), pkt) < 0) {
+	iface = get_iface(context, vlan_tag);
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
+#endif
```

Comment:
Have you tried to use DSA with VLAN?
I am not sure about the ordering of these two lines, just wondering should the `get_iface()` call be after the dsa one. It really depends how the VLAN master interface is setup (is it the DSA interface or one of the DSA slave interfaces).

Reply:
No, I did not tested it as my use case don't require and use VLANs. I just rely on the switch IC (KSZ8794) appended tag.

From the code in `drivers/ethernet/eth_mcux.c` it  seems like the VLAN related fields (TPID, VID) are only read (and the packet itself is not modified), so the tag will stay in place (appended to the end of the packet).

It looks to me that (by default) DSA shall be used when VLAN is disabled, otherwise (as you mentioned) there may be an issue with selected `iface`. This is not a problem, as user application may use registered `ctx->dsa_recv_cb` callback to just return 0 in `dsa_recv_set_iface()`, so the VLAN iface is used.